### ISSUE TITLE
Semantics of new header fields do not match established semantics

### ISSUE THREAD

**Issue by rmarx** [kazuho/draft-kazuho-httpbis-priority#27]
Within resource loading, there are several keywords that have highly-specific meanings. For example, blocking vs non-blocking (and then potentially render- vs parser blocking), progressive (typically used for progressive jpegs, other resources are sometimes referred to as "incremental"), deferred (javascript can be tagged as "async" or "defer" with specific and separate meanings), within service workers, "background" sync runs separately, etc. There are also several semantics that are currently not really expressible, e.g., preload (which are typically "blocking" resources but not until other blocking resources have been downloaded...)

The current proposal and comments from others on which fields to add, use names and values for the header fields that are very _similar_ to these semantics, but not entirely, with some strange and hard to define/explain edge cases, for example in issue #13. I feel that the solution taken in #13 to try and make things clearer is not a real solution. Another example previously mentioned was that if you would be viewing a single image in a tab, that image would be tagged as "document", which also feels weird. 

One of the goals of the header-based approach, as I see it, is to make it easier for "normal" web developers to view/adjust/change these values, set them in their own server configs, etc. Having close-but-not-really-identical semantics will make things *much* more difficult for people to figure out why things have certain values and how to adjust them to get desired behavior on the wire. (e.g., plenty of articles try to explain the difference between async and defer scripts, and then they would say: but for prioritization's sake, they are both "deferred").
___

I don't see easy solutions to this problem, but here are two that might have some merit (mainly concept, don't pin me down on exact names here):

**1. Do not re-use semantic-laden names**
`urgency=immediate&processing=incremental` 
or
`urgency=5&processing=incremental` 

vs

`urgency=blocking&progressive=1`

This avoids confusion while still being plenty clear on the intent. 

**2. Use exact mappings with much more metadata** 
assuming defaults of 
`aspects=non-blocking,non-incremental,immediate,body`

For a CSS file in the <head>:
`aspects=render-blocking,head`
For an image file above the fold:
`aspects=incremental,visible` (below the fold would loose 'visible')
For a pre-loaded Font in the <head>:
`aspects=preload,visible` (preload implies 'delayed' vs the default 'immediate')
For a defer-ed JS on the bottom of the HTML:
`aspects=defer` 

(probably made some mistakes here, but concept should be clear).
I generally like this approach a bit more and will open another issue to discuss that more in general. I like how the semantics are maintained and that it's trivial to add new "aspects" down the line should new browser loading primitives become available. The main downside is that it shifts -a lot- of complexity to the server implementers.

**Comment by kazuho** [comment]
>  __1. Do not re-use semantic-laden names__

I do not think that is a good idea. Consider prioritization between HTML and CSS. Chrome at the moment prioritizes HTML above CSS. We think that in most cases prioritizing CSS above HTML is better. And that is what Firefox does.

This example illustrates the problem of using numeric levels (or string that resembles numeric levels; e.g., "immediate") for prioritization. From a numeric signal, a server cannot tell if the request is for a HTML or a CSS, because, as stated, browsers might disagree on to which numeric level each resource should be mapped.

Using a semantic level resolves this problem. It is the premise of server controlling the prioritization (with the "help" of client-provided hints).

> __2. Use exact mappings with much more metadata__

I think that this is a good suggestion. It would be beneficial to transmit signals like initiator (the tag that initiated the request), or where the initiator resides in (e.g., head, body).

OTOH, I oppose to "replacing" urgency with such a signal. Because we do not have experience (and proof) that such signal works.

I think we should ship what we know that works (i.e. send the urgency levels that the browsers that use today, for building the prioiritization tree), _and_ consider having other signals, either as part of the draft or as a separate extension.

**Comment by rmarx** [comment]
This was partially resolved by PR #38. The other parts are in the discussions in #35 and #28. So I'm closing this issue.
